synfig-studio/src/gui/widgets/widget_timetrack.cpp

```diff
@@ -0,0 +1,825 @@
+	: waypoint_sd(*this),
+	  params_treeview(nullptr),
+	  is_rebuild_param_info_list_queued(false),
+	  is_update_param_list_geometries_queued(false)
+{
+	add_events(Gdk::BUTTON_PRESS_MASK | Gdk::BUTTON_RELEASE_MASK | Gdk::SCROLL_MASK | Gdk::POINTER_MOTION_MASK | Gdk::KEY_PRESS_MASK | Gdk::KEY_RELEASE_MASK);
+	set_can_focus(true);
+	time_plot_data->set_extra_time_margin(16/2);
+	setup_mouse_handler();
+
+	setup_adjustment();
+}
+
+Widget_Timetrack::~Widget_Timetrack()
+{
+	for (auto item : param_info_map) {
+		delete item.second;
+		item.second = nullptr;
+	}
+	teardown_params_view();
+	teardown_params_store();
+}
+
+bool Widget_Timetrack::set_params_view(Gtk::TreeView* treeview)
+{
+	teardown_params_view();
+	teardown_params_store();
+
+	params_treeview = nullptr;
+	params_store.reset();
+
+	if (!treeview)
+		return true;
+
+	Glib::RefPtr<LayerParamTreeStore> treestore = Glib::RefPtr<LayerParamTreeStore>::cast_dynamic( treeview->get_model() );
+	if (!treestore) {
+		synfig::error("TreeView must use model based on LayerParamTreeStore");
+		return false;
+	}
+	params_store = treestore;
+	params_treeview = treeview;
+	setup_params_store();
+	setup_params_view();
+
+	rebuild_param_info_list();
+
+	return true;
+}
+
+Gtk::TreeView* Widget_Timetrack::get_params_view() const
+{
+	return params_treeview;
+}
+
+Glib::RefPtr<LayerParamTreeStore> Widget_Timetrack::get_params_model() const
+{
+	return params_store;
+}
+
+bool Widget_Timetrack::use_canvas_view(etl::loose_handle<CanvasView> canvas_view)
+{
+	if (!canvas_view) {
+		synfig::error("No canvas_view for timetrack");
+		return false;
+	}
+	Gtk::TreeView* params_treeview = dynamic_cast<Gtk::TreeView*>(canvas_view->get_ext_widget("params"));
+	if (!params_treeview) {
+		synfig::error("Params treeview widget doesn't exist");
+		return false;
+	}
+
+	set_time_model(canvas_view->time_model());
+	set_canvas_interface(canvas_view->canvas_interface());
+	return set_params_view(params_treeview);
+}
+
+void Widget_Timetrack::delete_selected()
+{
+	std::lock_guard<std::mutex> lock(param_list_mutex);
+
+	// From CellRenderer_TimeTrack
+	synfigapp::Action::Handle action(synfigapp::Action::create("TimepointsDelete"));
+	if(!action)
+		return;
+
+	synfigapp::Action::ParamList param_list;
+	for (WaypointItem *wi : waypoint_sd.get_selected_items()) {
+		param_list.add("canvas", canvas_interface->get_canvas());
+		param_list.add("canvas_interface", canvas_interface);
+
+		const synfigapp::ValueDesc &value_desc = param_info_map[wi->path.to_string()]->get_value_desc();
+		if (value_desc.get_value_type() == synfig::type_canvas && !getenv("SYNFIG_SHOW_CANVAS_PARAM_WAYPOINTS")) {
+			param_list.add("addcanvas", value_desc.get_value().get(synfig::Canvas::Handle()));
+		} else {
+			param_list.add("addvaluedesc", value_desc);
+		}
+
+		param_list.add("addtime", wi->time_point.get_time());
+	}
+	action->set_param_list(param_list);
+	canvas_interface->get_instance()->perform_action(action);
+}
```

Comment:
@morevnaproject , could you try if a `queue_draw();` here solves your issue?

Reply:
This solves issue, but please see my next comment...

Reply:
I believe it's solved now.